apm821xx: wndr4700: fix uboot-env

neheb opened the pull request:
With nvmem-layout, these probe errors go away.

Switch to redundant-bool as it's the same but more widely used.

ping @chunkeey

neheb committed:
apm821xx: meraki-mx60: fix ubootenv definitions

Needed to avoid probe errors.

There are two partitions from 0-20000 and 80000-100000.

This is redundant-count and not regular u-boot,env

Add status = "disabled" as the u-boot env driver can't handle redundant
environments properly. As a result, fw_printenv ends up not working
right.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16618
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb committed:
apm821xx: wndr4700: fix uboot-env

With nvmem-layout, these probe errors go away.

Add status = "disabled" as the u-boot env driver can't handle redundant
environments properly. As a result, fw_printenv ends up not working
right.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16618
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb committed:
apm821xx: mybooklive: fix ubootenv probe

With nvmem-layout, these probe errors go away.

Add status = "disabled" as the u-boot env driver can't handle redundant
environments properly. As a result, fw_printenv ends up not working
right.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16618
Signed-off-by: Robert Marko <robimarko@gmail.com>

chunkeey commented:
From what I know, there's a real count and bool work differently... That's why there are two variants. In the end, this all depends which method the uboot uses to manage the uboot environment.

Question here: back when this was added, the redundant uboot feature didn't really work. Does it work now? Because the definitions are currently mostly only academical. (and there was also a [patch](https://lore.kernel.org/lkml/20221012155133.287-1-musashino.open@gmail.com/) necessary from @musashino205 to get it working, which wasn't merged).

neheb commented:
From reading https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/nvmem/layouts/u-boot-env.c?h=v6.12-rc2 , there's no difference. They both use U_BOOT_FORMAT_REDUNDANT. Only 3 formats are present.

It works. That patch was merged, at least locally: target/linux/generic/pending-6.6/802-nvmem-u-boot-env-align-endianness-of-crc32-values.patch

chunkeey commented:
> From reading https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/nvmem/layouts/u-boot-env.c?h=v6.12-rc2 , there's no difference. They both use U_BOOT_FORMAT_REDUNDANT. Only 3 formats are present.

Ok, then why not use the method that uboot on the devices uses? The binding document makes a clear distinction between -count and -bool: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/nvmem/layouts/u-boot,env.yaml

In case of the WNDR4700, this should be pretty easy to check since the uboot has to be updated in order to accommodate the never ending kernel expansions. 

That said, from what I know, with any "redundant" u-boot env support in the kernel isn't not implemented at all. The code doesn't know which of the redundant ubootenv is the currently active one. So I would wait with extracting values out of it, until this is implemented (it works with the existing userspace tools + they can also write to it)... Otherwise yeah, why not.

> It works. That patch was merged, at least locally: target/linux/generic/pending-6.6/802-nvmem-u-boot-env-align-endianness-of-crc32-values.patch

yes, locally. Are you fine with that or not?

neheb commented:
To be clear, the uboot env stuff is not used at all by OpenWrt. There's nothing in it that's useful.

This PR just fixes the definitions currently in dts as well as probe errors in dmesg.

musashino205 commented on target/linux/apm821xx/dts/wd-mybooklive.dts:
```diff
@@ -41,31 +41,33 @@
 
 				partition@1e000 {
-					compatible = "fixed-partitions";
 					reg = <0x0001e000 0x00002000>;
 					label = "u-boot-env";
+
+					compatible = "fixed-partitions";
```
`compatible` property should be placed at the first.

https://docs.kernel.org/devicetree/bindings/dts-coding-style.html#order-of-properties-in-device-node

neheb replied:
OK.

hauke commented:
As far as I understand the u-boot code the type of u-boot redundancy depends on the flash the u-boot environment is stored in, see https://source.denx.de/u-boot/u-boot/-/blob/master/tools/env/fw_env.c?ref_type=heads#L1487
For NAND and UBI it is incremental and for NOR bool.

The kernel driver does not support redundant partitions. If someone changes the MAC address in U-boot it could be that the kernel will still use the old value, because it uses the old partition. The user space tools handle this.

It would be very nice if someone could extend the uboot env driver in the kernel to also support redundancy.

The MAC address is not changing so often, so I think it should be ok to take the old value and I would favor the method of using the incomplete kernel driver.

hauke commented:
I think adding `u-boot-env-redundant` does not make much sense because we do not have support for it and the binding might look different. It is still nice to document this somehow in the device tree so we can later add it when we know how it will look.

neheb commented:
@hauke I propose a compromise.

Keep the u-boot,env definitions and disable CONFIG_NVMEM_U_BOOT_ENV=y as no device uses mac addresses from u-boot,env, which is the whole point of u-boot,env in dts at this time.

neheb committed:
apm821xx: disable NVMEM_U_BOOT_ENV

The main point of it currently is to extract mac addresses. That is not
being done as MAC addresses are elsewhere.

Disable it until it becomes more feature packed and there's an actual
use for it.

All devices already have config definitions. NVMEM prevents redundant
support as well as write support.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16618
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb commented:
@chunkeey added count back and disabled.

neheb commented:
anything else missing?

chunkeey commented on target/linux/apm821xx/dts/meraki-mx60.dts:
```diff
@@ -62,10 +62,30 @@
+					reg = <0x0 0x20000>;
+
+					nvmem-layout {
+						compatible = "u-boot,env-redundant-cunt";
```
Wot? I beg your pardon?! "u-boot,env-redundant-**cunt**" . So, you have acquired a MX60(W), is it too much to ask for a contributor to please test his own changes, right?

No no no, sorry I'm out. I can't take this abuse any longer.

neheb replied:
I must apologize for that. It's not intentional.

chunkeey commented on target/linux/apm821xx/dts/meraki-mx60.dts:
```diff
@@ -62,10 +62,30 @@
+					reg = <0x80000 0x20000>;
+
+					nvmem-layout {
+						compatible = "u-boot,env-redundant-cunt";
```
:disappointed:

chunkeey reviewed (changes requested):
is there a point in all this? :disappointed:

neheb commented:
The point here is to get rid of dmesg warnings about failing to probe.

neheb commented:
anything missing?

neheb commented:
ping?

raenye commented:
I manually installed `uboot-envtools` on the 6.12 build (without this patch) and now it works!

```
# apk add uboot-envtools
The following NEW packages will be installed:
  uboot-envtools
Need to download 17 KiB of packages.
After this operation, 104 KiB of additional disk space will be used.
(1/1) Installing uboot-envtools (2025.07-r1)
uboot-envtools-2025.07-r1.post-install: Executing script...
uboot-envtools-2025.07-r1.post-install: cfg016cb7
uboot-envtools-2025.07-r1.post-install: cfg026cb7
uboot-envtools-2025.07-r1.post-install: grep: /etc/fw_env.config: No such file or directory
OK: 82 MiB in 364 packages
```
```
# cat /etc/fw_env.config
/dev/mtd1 0x0 0x1000 0x1000 1
/dev/mtd1 0x1000 0x1000 0x1000 1
```
```
# fw_printenv
bootcmd=run boot_sata_script
bootdelay=5
baudrate=115200
loads_echo=
preboot=echo;echo Type "run flash_nfs" to mount root filesystem over NFS;echo
hostname=apollo3g
netdev=eth0
serverip=172.25.102.35
fresh_install=dhcp; run net_nfs
nfsargs=setenv bootargs root=/dev/nfs rw nfsroot=${serverip}:${rootpath}
ramargs=setenv bootargs root=/dev/ram rw
addip=setenv bootargs ${bootargs} ip=dhcp
addtty=setenv bootargs ${bootargs} console=ttyS0,${baudrate}
addmisc=setenv bootargs ${bootargs}
initrd_high=30000000
kernel_addr_r=1000000
fdt_addr_r=1800000
ramdisk_addr_r=1900000
hostname=apollo3g
bootfile=apollo3g/uImage
ramdisk_file=apollo3g/uRamdisk
sata_controller=sata1
load_boot=if run load_boot_file_1; then echo Loaded part 1; else echo Try part 2; run load_boot_file_2; fi
load_boot_file_1=sata init; ext2load sata 1:1 100000 /boot/boot.scr
load_boot_file_2=sata init; ext2load sata 1:2 100000 /boot/boot.scr
boot_sata_script=run load_boot; source 100000
rootpath=/nfs/debian_ppc/rootfs
flash_nfs=run nfsargs addip addtty addmisc;bootm ${kernel_addr} - ${fdt_addr}
net_nfs=tftp ${kernel_addr_r} ${bootfile}; tftp ${fdt_addr_r} ${fdt_file}; run nfsargs addip addtty addmisc;bootm ${kernel_addr_r} - ${fdt_addr_r}
net_self_load=tftp ${kernel_addr_r} ${bootfile};tftp ${fdt_addr_r} ${fdt_file};tftp ${ramdisk_addr_r} ${ramdisk_file};
net_self=run net_self_load;run ramargs addip addtty addmisc;bootm ${kernel_addr_r} ${ramdisk_addr_r} ${fdt_addr_r}
fdt_file=apollo3g/apollo3g.dtb
load=tftp 200000 apollo3g/u-boot.bin
update=protect off 0xFFFA0000 FFFFFFFF;era 0xFFFA0000 FFFFFFFF;cp.b ${fileaddr} 0xFFFA0000 ${filesize};setenv filesize;saveenv
upd=run load update
nload=tftp 200000 apollo3g/u-boot-nand.bin
nupdate=nand erase 0 100000;nand write 200000 0 100000;setenv filesize;saveenv
nupd=run nload nupdate
kernel_addr=fc000000
fdt_addr=fc1e0000
ramdisk_addr=fc200000
pciconfighost=1
pcie_mode=RP:RP
ethact=ppc_4xx_eth0
stdin=serial
stdout=serial
stderr=serial
ver=U-Boot 2009.08-svn65645 (Oct 08 2012 - 14:36:50)
filesize=1E
ipaddr=192.168.7.3
3g_self_test=0
ethaddr=00:90:A9:xx:xx:xx    (MAC address redacted)
do_vft=false
```

Note that the contents of `/etc/fw_env.config` is not the same as I wrote in the other thread, and after copying `fw_env,config` from the 6.12 machine to the 6.6.28 machine, `fw_printenv` worked fine there too.

neheb committed:
uboot-envtools: fix meraki mx60 definition

There are two redundant sections. One at 0x0 and the other at 0x80000.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16618
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb committed:
uboot-envtools: fix meraki mr24 definition

These two are redundant definitions according to dts. A value of 4 (CRC
no redundancy) makes no sense.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16618
Signed-off-by: Robert Marko <robimarko@gmail.com>

map-b commented:
Why your 6.12 build did not have uboot-envtools installed? And that 6.6.28 build is a bit old, some changes related has been merged since then.

neheb committed:
apm821xx: meraki-mr24: fix ubootenv definitions

Per the comments, this is not uboot,env but the redundant forms.

Placed under fixed-partition nodes in order to add status = "disabled".
The roots are needed for u-boot envtools to use.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16618
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb commented:
Updated PR to add status = "disabled"

raenye commented:
> Why your 6.12 build did not have uboot-envtools installed?

Probably I removed it from my config at some point; I guess it was after I found out it wasn't working fine.
(my previous, 6.6.87, config also has no `uboot-envtools`, but my 6.6.28 build still has it)

> And that 6.6.28 build is a bit old, some changes related has been merged since then.

Indeed, but this unit is not easy to access, so I prefer to update it less often.

> Updated PR to add status = "disabled"

Let me see if I understand what's going on here.

`/dev/mtd1` is the entire uboot-env, which consists of two identical copies? and these copies used to have DT entries, which are now not needed due to some improvements in `uboot-envtools`? and part of what these improvements do is to auto-create a proper `/etc/fw_env.config`?
And all of this would help to remove some manual MAC address initialization from `00_network` or so?

neheb commented:
another note:

```
case "$board" in
netgear,wndap620|\
netgear,wndap660)
        macaddr_add $(mtd_get_mac_ascii u-boot-env baseMAC) $(($PHYNBR + 1)) > /sys${DEVPATH}/macaddress
        ;;
esac
```

this cannot be handled by the uboot,env driver.

hauke commented:
There is a compile problem:
```
make[4]: Entering directory '/__w/openwrt/openwrt/openwrt/build_dir/target-powerpc-openwrt-linux-musl_musl/linux-apm821xx_nand/linux-6.6.97'
make[4]: Leaving directory '/__w/openwrt/openwrt/openwrt/build_dir/target-powerpc-openwrt-linux-musl_musl/linux-apm821xx_nand/linux-6.6.97'
make[4]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Error: ../dts/meraki-mr24.dts:81.6-26 Properties must precede subnodes
FATAL ERROR: Unable to parse input tree
make[4]: *** [Makefile:85: /__w/openwrt/openwrt/openwrt/build_dir/target-powerpc-openwrt-linux-musl_musl/linux-apm821xx_nand/image-meraki-mr24.dtb] Error 1
make[3]: *** [Makefile:27: dtb] Error 2
```

neheb commented:
@hauke should be fine now.

neheb commented:
@hauke rebased after 6.12 merge.
